File: 백준/Gold/1753. 최단경로/최단경로.py
import sys 
input = sys.stdin.readline
import heapq
INF = int(1e9)

V, E = map(int, input().split())
K = int(input())
graph = [[] for _ in range(V+1)]
dist = [INF]*(V+1)
for _ in range(E):
    u, v, w = map(int, input().split())
    graph[u].append((v, w))
    
def Dijkstra(start):
    queue = []
    heapq.heappush(queue, (0, start))
    dist[start] = 0
    
    while queue:
        d, cur = heapq.heappop(queue)
        if dist[cur] < d:
            continue
        for i in graph[cur]:
            cost = d + i[1]
            if cost < dist[i[0]]:
                dist[i[0]] = cost 
                heapq.heappush(queue, (cost, i[0]))
                
Dijkstra(K)

for i in range(1, V+1):
    if dist[i] == INF:
        print("INF")
    else:
        print(dist[i])

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

class Main {
    static int V, E, K;
    static ArrayList<ArrayList<Node>> graph = new ArrayList<>();
    static boolean[] visit;
    static int[] dist;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        K = Integer.parseInt(br.readLine());
        for (int i = 0; i < V+1; i++) {
            graph.add(new ArrayList<>());
        }
        visit = new boolean[V+1];
        dist = new int[V+1];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[K] = 0;

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine(), " ");
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            graph.get(u).add(new Node(v, w));
        }

        dijkstra();

        for (int i = 1; i < V+1; i++) {
            if (dist[i] == Integer.MAX_VALUE) System.out.println("INF");
            else System.out.println(dist[i]);
        }
    }

    static void dijkstra() {
        PriorityQueue<Node> pq = new PriorityQueue<>();
        pq.offer(new Node(K, 0));
        
        while (!pq.isEmpty()) {
            int cur = pq.poll().v;
            if (!visit[cur]) {
                visit[cur] = true;
                for (Node node : graph.get(cur)) {
                    if (dist[node.v] > dist[cur] + node.w) {
                        dist[node.v] = dist[cur] + node.w;
                        pq.add(new Node(node.v, dist[node.v]));
                    }
                }
            }
        }
    }

    static class Node implements Comparable<Node>{
        int v, w;

        Node(int v, int w) {
            this.v = v;
            this.w = w;
        }

        @Override
        public int compareTo(Node n) {
            return Integer.compare(this.w, n.w);
        }
    }
}
